src/__tests__/oplsNipals.test.js

```diff
@@ -0,0 +1,199 @@
+import { toBeDeepCloseTo } from 'jest-matcher-deep-close-to';
+import { getNumbers, getClasses } from 'ml-dataset-iris';
+import { METADATA } from 'ml-dataset-metadata';
+import { Matrix } from 'ml-matrix';
+
+import { oplsNipals } from '../oplsNipals';
+
+expect.extend({ toBeDeepCloseTo });
+
+const iris = getNumbers();
+const metadata = getClasses();
+const newM = new METADATA([metadata], { headers: ['iris'] });
+let numericValues = newM.get('iris', { format: 'matrix' }).values;
+
+describe('opls-nipals', () => {
+  it('test pls-nipals iris data', () => {
+    expect(iris).toHaveLength(150);
+    expect(metadata).toHaveLength(150);
+  });
+
+  it('test pls-nipals dataArray', () => {
+    let y = Matrix.from1DArray(150, 1, metadata);
+    let x = new Matrix(iris);
+
+    x = x.center('column').scale('column');
+    y = y.center('column').scale('column');
+
+    let model = oplsNipals(x, y);
+
+    expect(model.scoresXOrtho.to1DArray()).toHaveLength(150);
+  });
+
+  it('test opls-nipals with iris', () => {
+    let x = new Matrix(iris);
+    let y = numericValues.clone();
+    let model = oplsNipals(x, y);
+
+    expect(model.scoresXpred.getRow(0)).toBeDeepCloseTo([5.66576130528549]);
+    expect(model.scoresXpred.getRow(1)).toBeDeepCloseTo([5.357904200567914]);
+    expect(model.scoresXpred.getRow(148)).toBeDeepCloseTo([9.161531522704093]);
+    expect(model.scoresXpred.getRow(149)).toBeDeepCloseTo([8.539939678663997]);
+
+    expect(model.loadingsXpred.getRow(0)).toBeDeepCloseTo([
+      0.7533846125482135, 0.37991165000540095, 0.5175455876998517,
+      0.1697637836019074,
+    ]);
+
+    expect(model.weightsXPred.getColumn(0)).toBeDeepCloseTo([
+      0.7191635887943787, 0.3280487739174021, 0.5781304613979087,
+      0.20236823883090044,
+    ]);
+
+    expect(model.scoresXOrtho.getRow(0)).toBeDeepCloseTo([2.8553426763334433]);
+    expect(model.scoresXOrtho.getRow(1)).toBeDeepCloseTo([2.501798988578702]);
+    expect(model.scoresXOrtho.getRow(148)).toBeDeepCloseTo([-0.14717957843937]);
+    expect(model.scoresXOrtho.getRow(149)).toBeDeepCloseTo([-0.1097900904882]);
+
+    expect(model.loadingsXOrtho.getRow(0)).toBeDeepCloseTo([
+      1.893676431872939, 1.2763813703908802, 0.5781825736661366,
+      0.10014507134716187,
+    ]);
+
+    expect(model.weightsXOrtho.getRow(0)).toBeDeepCloseTo([
+      0.36913469592136616, 0.5594334971409369, -0.6535157770509538,
+      -0.35169712493733896,
+    ]);
+
+    expect(model.loadingsY.getRow(0)).toBeDeepCloseTo([0.14558532608468372]);
+  });
+
+  it('test opls-nipals with iris multi Y', () => {
+    let x = new Matrix(iris);
+    let y = new Matrix(createDummyY(metadata)).transpose();
+    let model = oplsNipals(x, y);
+    // > model$`Loadings X pred`
```

Comment:
here the R code is not useful because you don't include the first commands to get the model. You should complete the code so that we can copy/paste and get the results

Reply:
Sorry! My bad